### PR TITLE
build a stable dub version for dpl-docs

### PR DESCRIPTION
**Pull request by MartinNowak**
- generic shallow clone rule, either clones default branch (master)
  or a specific version

### CODE REVIEW

File: posix.mak [hunk 1]
```diff
@@ -234,15 +237,13 @@ $(DOC_OUTPUT_DIR)/dlangspec.pdf : dlangspec.dvi
...
-../%-${LATEST}/.cloned :
-	[ -d $(@D) ] || git clone ${GIT_HOME}/$* $(@D)/
-	if [ -d $(@D)/.git ]; then cd $(@D) && git checkout v${LATEST}; fi
-	touch $@
-
-# HEAD
+# dmd, druntime, dmd-2.066.1, dub-0.9.21
 ../%/.cloned :
-	[ -d $(@D) ] || git clone ${GIT_HOME}/$* $(@D)/
+ifneq ('',$(word 2,$(subst -, ,$*))) # clone a specific version
```
Comment (andralex) [inline posix.mak]: Are you sure this isn't `ifneq (,$(word 2,$(subst -, ,$*)))`

In other words, where do those quotes come about?

Comment (MartinNowak) [inline posix.mak]: Yeah, tried that before but $(word) returns an empty string which is different from an empty variable.
https://www.gnu.org/software/make/manual/html_node/Text-Functions.html#index-strip

Comment (andralex) [inline posix.mak]: My point is make strings don't have quotes. `''` is in fact a non-empty string consisting of two characters.

Comment (MartinNowak) [inline posix.mak]: You're right, that just can't work. I used two distinct rules instead.

File: posix.mak [hunk 2]
```diff
@@ -244,26 +247,25 @@ dlangspec.verbatim.txt : macros.ddoc verbatim.ddoc dlangspec-consolidated.d
 # Git clone rules
 ################################################################################
 
-# LATEST
-../%-${LATEST}/.cloned :
-	[ -d $(@D) ] || git clone ${GIT_HOME}/$* $(@D)/
-	if [ -d $(@D)/.git ]; then cd $(@D) && git checkout v${LATEST}; fi
+# dmd, druntime
+../%/.clone :
```
Comment (andralex) [inline posix.mak]: Why replace 'cloned' with 'clone'? The former was a lot more descriptive!

File: posix.mak [hunk 3]
```diff
@@ -244,26 +247,25 @@ dlangspec.verbatim.txt : macros.ddoc verbatim.ddoc dlangspec-consolidated.d
...
+	[ -d $(@D) ] || git clone --depth=1 ${GIT_HOME}/$* $(@D)/
 	touch $@
 
-# HEAD
-../%/.cloned :
-	[ -d $(@D) ] || git clone ${GIT_HOME}/$* $(@D)/
+# dmd-2.066.1, dub-0.9.21
+../%/.clonev :
```
Comment (andralex) [inline posix.mak]: So now we have clonev as well, and unexplained? How can we make this even more complicated?

File: posix.mak (hunk 2 above)
Comment (MartinNowak) [inline posix.mak]: Because `cloned` and `clonev` don't go well along.

File: posix.mak (hunk 3 above)
Comment (MartinNowak) [inline posix.mak]: > unexplained

The comment says `dmd-2.066.1, dub-0.9.21`.

> How can we make this even more complicated?

Well it's make and I tried to write a generic rule.
I solved this much simpler now.

Comment (andralex) [inline posix.mak]: Thank you!

**Comment by andralex** [comment]
Looks good, thanks!

**Comment by andralex** [comment]
After pulling this, `make clean` followed by `make` gets here:

    [ -d ../dub-0.9.22 ] || git clone --depth=1 https://github.com/D-Programming-Language/dub-0.9.22 ../dub-0.9.22/

This looks incorrect. Cloning over https (instead of git@ as we usually do) asks me for my username and password. Even after I enter them, it fails to authenticate me (I'm not sure why). But I think the actual command should attempt to clone `D-Programming-Language/dub`, not `D-Programming-Language/dub-0.9.22`. Right?

@MartinNowak, if the line above is in fact wrong, this is the second time untested code makes it in the makefiles. Asking for `make clean && make` to work is the absolute minimum anyone could expect. Could you please redo this and more thoroughly? We need to be extra careful (and much less keen) about changing makefiles because the larger the makefiles get the less robust they tend to become. Thanks.

**Comment by andralex** [comment]
After I run the corrected clone command, I get to this command:

    cd ../dub-0.9.22; DC=/Users/aalexandre/code/d/dlang.org/.stable_dmd-2.066.1/dmd2/osx/bin/dmd ./build.sh

with the result:

	Generating version file...
	fatal: No names found, cannot describe anything.
	Running /Users/aalexandre/code/d/dlang.org/.stable_dmd-2.066.1/dmd2/osx/bin/dmd...
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(859): Error: found 'nothrow' when expecting '{'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(861): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(862): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(863): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(864): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(865): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(866): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(867): Error: mismatched number of curly brackets
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(871): Error: asm statements must end in ';'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(874): Error: found 'private' instead of statement
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(884): Error: no identifier for declarator add
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(887): Error: no identifier for declarator usDone
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(887): Error: Declaration expected, not ':'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(894): Error: Declaration expected, not '('
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(896): Error: Declaration expected, not 'foreach'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(896): Error: Declaration expected, not '0'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(901): Error: no identifier for declarator __fhnd_info[fd]
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(901): Error: Declaration expected, not '='
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(902): Error: Declaration expected, not 'return'
	/Users/aalexandre/code/d/druntime/import/core/stdc/stdio.d(904): Error: unrecognized declaration
	/Users/aalexandre/code/d/druntime/import/core/thread.d(179): Error: found 'nothrow' when expecting '{'

What am I doing wrong? Thanks.

**Comment by MartinNowak** [comment]
Probably nothing, looks like the OSX release zip comes with an incorrect dmd.conf.
It shouldn't pick up your local druntime.

**Comment by MartinNowak** [comment]
> This looks incorrect. Cloning over https (instead of git@ as we usually do) asks me for my username and password.

https is correct, the url is wrong, cloning over ssh ask everyone bug DPL members for a password
https://help.github.com/articles/which-remote-url-should-i-use/#cloning-with-https-recommended

> @MartinNowak, if the line above is in fact wrong, this is the second time untested code makes it in the makefiles. Asking for make clean && make to work is the absolute minimum anyone could expect.

As if I hadn't done that, careful with accusations.
We're using different make versions, therefor this slipped through.
http://stackoverflow.com/questions/11455182/when-multiple-pattern-rules-match-a-target

fixed by #746

BTW, I'm already trying to reach @braddr for quite a while to get dlang.org into the auto-tester.
The documentation breaks way too often without CI.
Maybe you know how to reach him. https://github.com/braddr/d-tester/issues/41

**Comment by MartinNowak** [comment]
As for the second part, I get this.

```
cd ../dub-0.9.22; DC=/Users/vagrant/dlang.org/.stable_dmd-2.066.1/dmd2/osx/bin/dmd ./build.sh
Generating version file...
Running /Users/vagrant/dlang.org/.stable_dmd-2.066.1/dmd2/osx/bin/dmd...
DUB has been built as bin/dub.
```

This looks wrong in your output, have you cloned the tagged version?
```
fatal: No names found, cannot describe anything.
```

Don't know why your stable dmd picks up an different dmd.conf.
Can you please run `.stable_dmd-2.066.1/dmd2/osx/bin/dmd -c -v nonexisting.d` in the dlang.org folder. The config line should be like this.
```
config    .stable_dmd-2.066.1/dmd2/osx/bin/dmd.conf
```

**Comment by andralex** [comment]
@MartinNowak: you are right, apologies for my tone.

On cloning: shall we then use a unified method for all similar cases? Whichever is best for our setup.

On config: I ran `.stable_dmd-2.066.1/dmd2/osx/bin/dmd -c -v nonexisting.d`

    binary    .stable_dmd-2.066.1/dmd2/osx/bin/dmd
    version   v2.066.1
    config    /Users/aalexandre/dmd.conf
    parse     nonexisting
    Error: cannot read file nonexisting.d` 

So it picked up my dmd.conf, which I use for specific purposes for my machine. That seems to be the matter. The file `.stable_dmd-2.066.1/dmd2/osx/bin/dmd.conf` does exist - how come the one in my $HOME takes precedence?

I'll reach out to Brad.

**Comment by MartinNowak** [comment]
> On cloning: shall we then use a unified method for all similar cases? Whichever is best for our setup.

That didn't work out before, the easiest fix here is to change the order, see #746.

**Comment by MartinNowak** [comment]
> how come the one in my $HOME takes precedence?

Walter must have had a bad day when he decided about the search order for dmd.conf.
https://github.com/D-Programming-Language/dmd/blob/d7ef79918c905f337ceb738c03925204cd291a62/src/inifile.c#L51

**Comment by MartinNowak** [comment]
See https://github.com/D-Programming-Language/dmd/pull/4256

**Pull request merged**